src/Kriti/Eval.hs

```diff
@@ -12,82 +13,102 @@ import qualified Data.Text.Encoding as TE
+  serialize :: EvalError -> SerializedError
+  serialize (InvalidPath _ term path) = SerializedError {_code = InvalidPathCode, _message = "\"" <> renderVect path <> "\"", _span = locate term}
+  serialize (TypeError _ term msg) = SerializedError {_code = TypeErrorCode, _message = msg, _span = locate term}
+  serialize (RangeError _ term) = SerializedError {_code = RangeErrorCode, _message = "Can only range over an array", _span = locate term}
+
+type Ctxt = (B.ByteString, Compat.Object J.Value)
 
 getSourcePos :: EvalError -> Span
-getSourcePos (InvalidPath pos _) = pos
-getSourcePos (TypeError pos _) = pos
-getSourcePos (RangeError pos) = pos
+getSourcePos (InvalidPath _ pos _) = locate pos
+getSourcePos (TypeError _ term _) = locate term
+getSourcePos (RangeError _ pos) = locate pos
 
-evalPath :: J.Value -> V.Vector (Accessor) -> ExceptT EvalError (Reader Ctxt) J.Value
-evalPath ctx path =
-  let step :: Monad m => J.Value -> (Accessor) -> ExceptT EvalError m J.Value
-      step (J.Object o) (Obj sp k) = maybe (throwError $ InvalidPath sp path) pure $ Compat.lookup k o
-      step (J.Array xs) (Arr sp i) = maybe (throwError $ InvalidPath sp path) pure $ xs V.!? i
-      -- TODO: Should we extend this error message with the local Context?
-      step _ (Obj sp _) = throwError $ TypeError sp "Expected object"
-      step _ (Arr sp _) = throwError $ TypeError sp "Expected array"
+evalPath :: Span -> J.Value -> V.Vector (Accessor) -> ExceptT EvalError (Reader Ctxt) J.Value
+evalPath sp ctx path = do
+  src <- asks fst
+  let step :: Monad m => J.Value -> Accessor -> ExceptT EvalError m J.Value
+      step (J.Object o) (Obj _ k _) = maybe (throwError $ InvalidPath src sp path) pure $ Compat.lookup k o
+      step (J.Array xs) (Arr _ i) = maybe (throwError $ InvalidPath src sp path) pure $ xs V.!? i
+      step _ (Obj _ _ _) = throwError $ TypeError src sp "Expected object"
+      step _ (Arr _ _) = throwError $ TypeError src sp "Expected array"
    in foldlM step ctx path
 
 isString :: J.Value -> Bool
 isString J.String {} = True
 isString _ = False
 
-runEval :: ValueExt -> [(T.Text, J.Value)] -> Either EvalError J.Value
-runEval template source =
+runEval :: B.ByteString -> ValueExt -> [(T.Text, J.Value)] -> Either EvalError J.Value
+runEval src template source =
   let ctx = Compat.fromList source
-   in runReader (runExceptT (eval template)) ctx
+   in runReader (runExceptT (eval template)) (src, ctx)
 
-serializeType :: J.Value -> T.Text
-serializeType J.Object {} = "Object"
-serializeType J.Array {} = "Array"
-serializeType J.String {} = "String"
-serializeType J.Number {} = "Number"
-serializeType J.Bool {} = "Boolean"
-serializeType J.Null = "Null"
+typoOfJSON :: J.Value -> T.Text
+typoOfJSON J.Object {} = "Object"
+typoOfJSON J.Array {} = "Array"
+typoOfJSON J.String {} = "String"
+typoOfJSON J.Number {} = "Number"
+typoOfJSON J.Bool {} = "Boolean"
+typoOfJSON J.Null = "Null"
 
 eval :: ValueExt -> ExceptT EvalError (Reader Ctxt) J.Value
```

Comment:
I haven't gotten a chance to survey the codebase lately, but depending on how common this transformer is it might make sense to just pack it up into a `KritiM` newtype at some point.

I figure most interpreters/evaluators end up with some sort of environmental context (`ReaderT`) and error handler (`ExceptT`) anyway, so it seems like it would make sense.

Reply:
Yeah this transformer stack is extremely common for intepreters. i didn't bother with a newtype cause this interpreter is so small. We only have 2 functions with that type. I figured I would rethink this decision later if the interpreter starts to grow in complexity.